### ISSUE TITLE
Enum case sensitivity

### ISSUE THREAD

**Issue by iamdanfox** [palantir/conjure#193]
## Motivation

From discussion on https://github.com/palantir/conjure-java/issues/134, it appears that the case sensitivity of Conjure enums is not defined precisely enough in the wire format spec. @markelliot provided context that early adopters may still be relying on the case insensitivity of enums. This needs to be defined precisely so that new languages (e.g. Rust) can be implemented correctly and precise positive and negative test cases can be added to the conjure-verification server.

Given an enum definition:

```yml
Direction:
  values:
  - UP
  - DOWN
  - LEFT
  - RIGHT
```

Should a message of `{"direction": "up"}` be accepted by a Conjure server as equivalent to `{"direction": "UP"}`? Also should this lowercase presentation of `"up"` be preserved when re-serialized?

## Option 1: lenient deserialization, round-trip preserves case

The enum values `RIGHT`, `right`, `rIgHt` should all deserialize without error and be equivalent to a canonical enum value of `Direction.RIGHT`.  Re-serializing these values should produce the same input strings: `RIGHT`, `right` and `rIgHt`.

Clearly this option minimizes the chance of servers rejecting an input.  However, it does require all Conjure language implementations to store the original casing information for each enum to enable re-serialization, requiring slightly higher memory overhead for each enum instance. This is not currently supported by conjure-java and would be a significant undertaking to add to conjure-typescript (as typescript enums are currently pure strings and there isn't a convenient way of overriding equality in typescript so that a switch statement would match both `rIgHt` and `RIGHT`).

## Option 2: lenient deserialization, no round-trip 

This is conjure-java's behaviour, as of 2.6.2. 

```java
    @Test
    public void testEnumCasingDeserializationInvariantToInputCase() throws Exception {
        assertThat(mapper.readValue("\"ONE\"", EnumExample.class)).isEqualTo(EnumExample.ONE);
        assertThat(mapper.readValue("\"one\"", EnumExample.class)).isEqualTo(EnumExample.ONE);
        assertThat(mapper.readValue("\"onE\"", EnumExample.class)).isEqualTo(EnumExample.ONE);
        assertThat(mapper.readValue("\"oNE\"", EnumExample.class)).isEqualTo(EnumExample.ONE);
    }

    @Test
    public void enumSerializationDoesntPreserveCase() throws Exception {
        assertThat(mapper.writeValueAsString(mapper.readValue("\"one\"", EnumExample.class))).isEqualTo("\"ONE\"");
        assertThat(mapper.writeValueAsString(mapper.readValue("\"ONE\"", EnumExample.class))).isEqualTo("\"ONE\"");
        assertThat(mapper.writeValueAsString(mapper.readValue("\"onE\"", EnumExample.class))).isEqualTo("\"ONE\"");
    }
```

Notably conjure-typescript does _not_ support this lenient deserialization as there isn't a serialization layer yet (https://github.com/palantir/conjure-typescript/issues/48).  As such, I think we can be sure that no conjure-typescript clients tolerate receiving lowercase enums from a server, even if they may somehow still be sending them.

## Option 3: strict deserialization

A request containing `right` would be considered invalid, and would be rejected by the server. Clients could be certain that all enum responses would be uppercase.  This minimizes the amount of memory necessary to store enums internally and also makes new language implementations simpler.

**Comment by cakofony** [comment]
I'm in favor of option 3 as well -- I don't like complicated things.

**Comment by cakofony** [comment]
Should we differentiate between unknown and unknowable?
The value `SOME_ENUM_VALUE` could potentially match across api versions (unknown) but `Some Enum 💯  value` is unknowable and we are confident that it is not, nor ever will be part of the enum.

I would prefer not to deviate handling enum and union key interpretation. It would be odd if one were case senseitive and the other not. Do we have any prior art in the conjure spec for case insensitivity?

**Comment by iamdanfox** [comment]
I'd like to close this out and go with option (3), as implemented https://github.com/palantir/conjure-java/pull/143 - we've been shoring up the spec throughout the open-sourcing process and I think this is just one more under-defined topic to iron out.  

I'm confident that this won't cause lots of disruption for users because TypeScript users have never been able to switch on lowercase enum values, so I don't think they've ever been used.

**Comment by markelliot** [comment]
Will closing this out also involve updating all implementations to explicitly treat received lowercase values with an explicit, differentiated error from unknown but uppercase values?

**Comment by sfackler** [comment]
Yeah, I think it makes sense to limit the unknown variant handling to strings which could actually be valid but unknown variants. That's just an ASCII SCREAMING_SNAKE_CASE check, right?

**Comment by sfackler** [comment]
The same goes for unknown union variant names too, I suppose.

**Comment by sfackler** [comment]
Oh yeah - I meant that we should make sure that unknown union variant names are eligible to be union variant names. The check would be for snakeCase for those I believe rather than SCREAMING_SNAKE_CASE.

**Comment by cakofony** [comment]
lowerCamelCase

**Comment by sfackler** [comment]
Er yeah, that >_>

**Comment by iamdanfox** [comment]
For context, @markelliot raised concerns in a slack thread that the newly imposed enum strictness imposes an unfair burden on any existing user-facing services we have that are relying on the old lenient behaviour.

https://github.com/palantir/conjure-java/pull/232 is a feature flag which would allow existing services to keep the old behaviour around.
